src/fast_forward.rs

```diff
@@ -0,0 +1,505 @@
+use crate::error::{Error, Result};
+use chia_protocol::Bytes32;
+use chia_protocol::Coin;
+use clvm_traits::{FromClvm, ToClvm};
+use clvm_utils::CurriedProgram;
+use clvm_utils::{tree_hash, tree_hash_atom, tree_hash_pair};
+use clvmr::allocator::{Allocator, NodePtr};
+use hex_literal::hex;
+
+#[derive(FromClvm, ToClvm, Debug)]
+#[clvm(tuple)]
+pub struct SingletonStruct {
+    pub mod_hash: Bytes32,
+    pub launcher_id: Bytes32,
+    pub launcher_puzzle_hash: Bytes32,
+}
+
+#[derive(FromClvm, ToClvm, Debug)]
+#[clvm(curried_args)]
+pub struct SingletonArgs {
+    pub singleton_struct: SingletonStruct,
+    pub inner_puzzle: NodePtr,
+}
+
+#[derive(FromClvm, ToClvm, Debug)]
+#[clvm(proper_list)]
+pub struct LineageProof {
+    pub parent_parent_coin_id: Bytes32,
+    pub parent_inner_puzzle_hash: Bytes32,
+    pub parent_amount: u64,
+}
+
+#[derive(FromClvm, ToClvm, Debug)]
+#[clvm(proper_list)]
+pub struct SingletonSolution {
+    pub lineage_proof: LineageProof,
+    pub amount: u64,
+    pub inner_solution: NodePtr,
+}
+
+// TODO: replace this with a generic function to compute the hash of curried
+// puzzles
+fn curry_single_arg(arg_hash: [u8; 32], rest: [u8; 32]) -> [u8; 32] {
+    tree_hash_pair(
+        tree_hash_atom(&[4]),
```

Comment:
Maybe use a named constant for `4` so it's more clear what operators are involved.